clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2784,16 +2784,80 @@ class SyclOptReportCreator : public SyclKernelFieldHandler {
   }
 };
 
-static CXXMethodDecl *getOperatorParens(const CXXRecordDecl *Rec) {
-  for (auto *MD : Rec->methods()) {
-    if (MD->getOverloadedOperator() == OO_Call)
-      return MD;
+// This function traverses the static call graph from the root of the kernel
+// (e.g. “kernel_parallel_for”) and returns the version of “operator()()” that
+// is called by kernelFunc(). There will only be one call to kernelFunc() in
+// that call graph because the DPC++ headers are structured such that the user’s
+// kernel function is only called once. This ensures that the correct
+// “operator()()” function call is returned, when a named function object used
+// to define a kernel has more than one “operator()()” calls defined in it. For
+// example, in the code below, 'operator()(sycl::id<1> id)' is returned based on
+// the 'parallel_for' invocation which takes a 'sycl::range<1>(16)' argument.
+//   class MyKernel {
+//    public:
+//      void operator()() const {
+//        // code
+//      }
+//
+//      [[intel::reqd_sub_group_size(4)]] void operator()(sycl::id<1> id) const
+//      {
+//        // code
+//      }
+//    };
+//
+//    int main() {
+//
+//    Q.submit([&](sycl::handler& cgh) {
+//      MyKernel kernelFunctorObject;
+//      cgh.parallel_for(sycl::range<1>(16), kernelFunctorObject);
+//    });
+//      return 0;
+//    }
+
+static CXXMethodDecl *
+getCallOperatorInvokedFromKernel(const CXXRecordDecl *KernelFuncObjType,
+                                 FunctionDecl *KernelCallerFunc,
+                                 Sema &SemaRef) {
+
+  CallGraph SYCLCG;
+  SYCLCG.addToCallGraph(SemaRef.getASTContext().getTranslationUnitDecl());
+
+  // This code returns the 'lambda' call operator.
+  if (KernelFuncObjType->isLambda()) {
```

Comment:
This function was originally used to return the call operator of lambda functions as well.

If the kernel is defined as a lambda, we can simply find the call operator without having to traverse the call graph from the kernel root.

Also, when we reach this function via `isESIMDKernelType()` from `Sema::CheckSYCLKernelCall`,
the `KernelCallerFunc()` is still a template and does not have a definition generated for it.
So it is not possible to set a call graph when the kernel root doesn't have a definition/body associated with it.

Reply:
Maybe should involve an ESIMD reviewer at some point.

Reply:
> If there is a cleaner way to handle this case, please let me know.

Probably move some diagnosing to `Sema::MarkDevices". There is already some happening, also via callgraph.